UIPS/UIP-0123.md

```diff
@@ -17,27 +17,26 @@ Implementing this hint would remove the need to manually implement such checks w
 
 ## Motivation
 
-Some Nock programs such as the Hoon compiler must run complex recursive functions over arbitrary user input. Algorithms which rely on memoization to be feasible can leverage the `%memo` hint supported by both Vere and Ares.
+The `%memo` hint, supported by both Vere and Ares, allows algorithms to improve performance using memoization. The underlying cache for the hint is, in general, a map from subject-formula pair to result.
 
-However, if it is necessary to deduce non-termination and crash with an error, the necessary state in the subject prevents the use of the %memo hint. While this is not the only reason for custom memoizing jets in the Hoon compiler, it is a major contributor. Non-termination and crashes are equivalent in Nock, and an interpreter choosing to report a crash is reporting that the computation would not terminate if continued.
+Some Nock programs, such as the Hoon compiler, must run complex recursive functions over arbitrary user input. These programs need memoization to be feasible. However, they also need to crash with an error when they detect non-termination. Since the programs store the state used to deduce non-termination in the subject, they're unable to use the `%memo` hint. This is a major contributor to the custom memoization jets for the Hoon compiler functions, which manually punch out holes in the subject before caching.
```

Comment:
```suggestion
Some Nock programs, such as the Hoon compiler, must run complex recursive functions over arbitrary user input. These programs need memoization to be feasible. However, they also need to crash with an error when they detect non-termination. Since the programs store the state used to deduce non-termination in the subject, and this state changes each iteration, they're unable to use the `%memo` hint. This is a major contributor to the custom memoization jets for the Hoon compiler functions, which manually punch out holes in the subject before caching.
```